### DISCUSSION TITLE
How do I connect multiple to an array field?

### DISCUSSION THREAD

**Discussion by shivamsingha** [prisma/prisma#4709] [Q&A] [answered]
Sorry I couldn't make the title less dumb but I'll try my best to explain.
So I have a model that looks kinda like this. Now I want to get the array of Items from **Profile.cart**, add them to a new **Order**, and add the **Order** to **Profile.orders[]**.

Edit: I keep getting this TS error
```js
Type '{ items: { connect: { id: number; }[]; }; }' is not assignable to type 
'OrderCreateWithoutProfileInput | OrderCreateWithoutProfileInput[] | undefined'.

Type '{ items: { connect: { id: number; }[]; }; }' is not assignable to type 'undefined'.ts(2322)
```
```prisma
model Item {
  id          Int      @id @default(autoincrement())
  Order       Order?   @relation(fields: [orderId], references: [id])
  orderId     Int?
  Profile     Profile? @relation(fields: [profileId], references: [id])
  profileId   Int?
}

model Order {
  id             Int      @id @default(autoincrement())
  items          Item[]
  User           User?    @relation(fields: [userId], references: [id])
  userId         Int?
  Profile        Profile? @relation(fields: [profileId], references: [id])
  profileId      Int?
}

model Profile {
  id      Int     @id @default(autoincrement())
  cart    Item[]
  orders  Order[]
  user    User    @relation(fields: [userId], references: [id])
  userId  Int     @unique
}

model User {
  id          Int           @id @default(autoincrement())
  profile     Profile?
  Order       Order[]
}
```
Now I'm stuck figuring out how to do it. I'm trying something like this.

```js
prisma.user
    .findUnique({
      where: { id: userId },
      select: {
        profile: {
          select: {
            cart: {
              select: {
                id: true
              }
            }
          }
        }
      }
    })
    .then((cart) =>
      prisma.user.update({
        where: { id: userId },
        data: {
          profile: {
            upsert: {
              create: {
                orders: {
                  create: {
                    items: {
                      connect: { 
                        id: 1                 // id: [array of numbers] also throws the same TS error mentioned above
                      }                       // ?? How do I do this? 
                    }
                  }, 
                  update: {...}
                }
              }
            }
          }
        }
      })
    )
```

I read the docs but I'm still unable to figure it out. Please help.
Thank you.

**Comment by ryands17** [comment, marked as answer]
Hey @shivamsingha 👋 
You can directly pass an array of id's inside `connect` to add multiple items. So the above query would be as follows:

```ts
prisma.user
    .findUnique({
      where: { id: userId },
      select: {
        profile: {
          select: {
            cart: {
              select: {
                id: true,
              },
            },
          },
        },
      },
    })
    .then(cart =>
      prisma.user.update({
        where: { id: userId },
        data: {
          profile: {
            upsert: {
              create: {
                orders: {
                  create: {
                    items: {
                      connect:
                        cart?.profile?.cart.map(c => ({ id: c.id })) || [],
                    },
                  },
                },
              },
              update: {},
            },
          },
        },
      })
    )
```

You can read more about that [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#nested-writes) under the `User` and `Post` example :)

**Reply by shivamsingha** [reply]
Thank you for your help. But I'm having a problem. I'm getting TS errors.
```js
Type '{ items: { connect: { id: number; }[]; }; }' is not assignable to type 'OrderCreateWithoutProfileInput | OrderCreateWithoutProfileInput[] | undefined'.
Type '{ items: { connect: { id: number; }[]; }; }' is not assignable to type 'undefined'.ts(2322)
```

**Reply by ryands17** [reply]
I'm not getting any TypeScript issue here:

<img width="853" alt="Screenshot 2020-12-18 at 6 52 59 PM" src="https://user-images.githubusercontent.com/19697099/102619320-4844c900-4162-11eb-876a-42170551b2dd.png">

What version of Prisma Client and CLI are you using?

**Reply by shivamsingha** [reply]
version 2.13, both client and cli

**Reply by ryands17** [reply]
Could you share a or code snippet as I am unable to reproduce this error that you're getting.

**Reply by shivamsingha** [reply]
[https://gist.github.com/shivamsingha/84ff765b2df79cbfe71d1bf986d507c2](https://gist.github.com/shivamsingha/84ff765b2df79cbfe71d1bf986d507c2)

Would this be enough?

**Reply by ryands17** [reply]
@shivamsingha In this case, I took the relevant part of your schema as follows:

```prisma
model Item {
  id        Int      @id @default(autoincrement())
  Order     Order?   @relation(fields: [orderId], references: [id])
  orderId   Int?
  Profile   Profile? @relation(fields: [profileId], references: [id])
  profileId Int?
}

model Order {
  id        Int      @id @default(autoincrement())
  items     Item[]
  User      User?    @relation(fields: [userId], references: [id])
  userId    Int?
  Profile   Profile? @relation(fields: [profileId], references: [id])
  profileId Int?
}

model Profile {
  id     Int     @id @default(autoincrement())
  cart   Item[]
  orders Order[]
  user   User    @relation(fields: [userId], references: [id])
  userId Int     @unique
}

model User {
  id      Int      @id @default(autoincrement())
  profile Profile?
  Order   Order[]
}
```

Then I ran the following commands (version _2.13.0_ of `@prisma/client` and `@prisma/cli`.

```
npx prisma migrate dev --create-only --preview-feature
npx prisma migrate dev --preview-feature
```

I created some dummy data for running the query like this:

```ts
await prisma.user.create({
    data: { profile: { create: { cart: { create: [{}, {}, {}, {}] } } } },
})
```

And then ran the following query from your code:

```ts
return prisma.user
.findUnique({
  where: { id: userId },
  select: {
    profile: {
      select: {
        cart: {
          select: {
            id: true,
          },
        },
      },
    },
  },
})
.then(cart =>
  prisma.user
    .update({
      where: { id: userId },
      data: {
        profile: {
          upsert: {
            create: {
              orders: {
                create: {
                  items: {
                    connect:
                      cart?.profile?.cart.map(c => ({ id: c.id })) || [],
                  },
                },
              },
            },
            update: {
              orders: {
                create: {
                  items: {
                    connect:
                      cart?.profile?.cart.map(c => ({ id: c.id })) || [],
                  },
                },
              },
            },
          },
        },
      },
      select: {
        profile: {
          select: {
            orders: true,
          },
        },
      },
    })
    .then(orders => console.log(JSON.stringify(orders, null, 2)))
)
```

Everything works as expected and here's the result of completion:
![image](https://user-images.githubusercontent.com/19697099/102765882-43bb2300-43a3-11eb-80dd-0bfb602ad1fd.png)

Let me know if the above steps work for you.

**Reply by shivamsingha** [reply]
Still getting the same error. Is it something to do with Typescript? 
I'm using typescript 3.9.7 if that's relevant.

**Reply by ryands17** [reply]
Yeah I'm using TS 4.1.3, so you can try updating and check. Also try it on a clean database with just the inserted data like I have done above and see if it works. If it still doesn't, then I would need to check your entire flow to see what's the issue.

**Reply by shivamsingha** [reply]
Still getting the error in 4.1.3

**Comment by wambugucoder** [comment]
Could i have a look at your typescript version and the prisma version

**Reply by shivamsingha** [reply]
Prisma 2.13.0 & Typescript 4.1.3

**Reply by wambugucoder** [reply]
@shivamsingha  I would reccomend  you try reinstalling all your dependencies  and if that doesnt seem to work,  you could return back to an  earlier version of those dependencies. 
Do update me on how that goes.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
